src/Microsoft.AspNetCore.Razor.Evolution/LargeTextRazorSourceDocument.cs

```diff
@@ -0,0 +1,136 @@
+
+        public LargeTextRazorSourceDocument(StreamReader reader, int chunkMaxLength, Encoding encoding, string filename)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            _chunkMaxLength = chunkMaxLength;
+            Encoding = encoding;
+            Filename = filename;
+
+            ReadChunks(reader, _chunkMaxLength, out _length, out _chunks);
+            _lines = new DefaultRazorSourceLineCollection(this);
+        }
+
+        public override char this[int position]
+        {
+            get
+            {
+                var chunkIndex = position / _chunkMaxLength;
+                var insideChunkPosition = position - chunkIndex * _chunkMaxLength;
+
+                return _chunks[chunkIndex][insideChunkPosition];
+            }
+        }
+
+        public override Encoding Encoding { get; }
+
+        public override string Filename { get; }
+
+        public override int Length => _length;
+
+        public override RazorSourceLineCollection Lines => _lines;
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+        {
+            if (destination == null)
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+
+            if (sourceIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(sourceIndex));
+            }
+
+            if (destinationIndex < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(destinationIndex));
+            }
+
+            if (count < 0 || count > Length - sourceIndex || count > destination.Length - destinationIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (count == 0)
+            {
+                return;
+            }
+
+            var chunkIndex = sourceIndex / _chunkMaxLength;
+            var insideChunkPosition = sourceIndex - chunkIndex * _chunkMaxLength;
+            var remaining = count;
+            var currentDestIndex = destinationIndex;
+
+            while (remaining > 0)
+            {
+                var toCopy = Math.Min(remaining, _chunkMaxLength - insideChunkPosition);
+                Array.Copy(_chunks[chunkIndex], insideChunkPosition, destination, currentDestIndex, toCopy);
+
+                remaining -= toCopy;
+                currentDestIndex += toCopy;
+                chunkIndex++;
+                insideChunkPosition = 0;
+            }
+        }
+
+        private static void ReadChunks(StreamReader reader, int chunkMaxLength, out int length, out List<char[]> chunks)
+        {
+            length = 0;
+            chunks = new List<char[]>();
+
+            for (;;)
```

Comment:
It would be nice to avoid a potentially infinite loop here, we get a bit nervous about `for(;;)` or `while(true)`. 

This can be simplified because `ReadBlock` will only read fewer than `count` characters when it's reached the end. https://msdn.microsoft.com/en-us/library/system.io.streamreader.readblock(v=vs.110).aspx#Anchor_2

```

char[] chunk;
int read;

do
{
    chunk = new char[chunkMaxLength];
    read = reader.ReadBlock(chunk, index, chunkMaxLength);
    length += read;
    if (read > 0)
    {
        chunks.Add(chunk);
    }
}
while (read == chunkMaxLength);
```